### PR TITLE
[SC-42] August 16 spell

### PR DESCRIPTION
**Pull request by hexonaut**
This spell will update the pool implementation to Aave V3.0.2, unfreeze the sDAI market and change the DAI market interest rate strategy back to target the DSR. The Maker main spell will also transfer all remaining ownership to the Spark Proxy BEFORE executing this spell.

This also contains a large amount of updates to the `aave-helpers` repository to drastically improve e2e test coverage.

Tenderly fork of mainnet spell: https://dashboard.tenderly.co/shared/fork/fb8ad43a-c2bd-4d9d-8153-f6a7d539c399/transactions (Note this also calls `transferOwnership()` before executing the spark spell which is in the Maker main spell.)

### CODE REVIEW

**Comment by hexonaut** [comment]
The spell itself is ready for review, but work is still being done in `aave-helpers` to have more robust e2e testing for the Pool implementation upgrade.

File: src/proposals/20230816/SparkGoerli_20230816.t.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                ilk:                SPARK_ILK,
+                baseRateConversion: 1e27,
+                borrowSpread:       0,
+                supplySpread:       0,
+                maxRate:            0.75e27,
+                performanceBonus:   0
+            })
+        );
+        assertEq(_findReserveConfigBySymbol(configsBefore, 'sDAI').isFrozen, true);
+
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20230816/SparkGoerli_20230816.t.sol:71]: done

File: src/proposals/20230816/SparkEthereum_20230816.sol [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import { SparkPayloadEthereum, IEngine, Rates, EngineFlags } from '../../SparkPayloadEthereum.sol';
+
+import { IPoolAddressesProvider } from 'aave-v3-core/contracts/interfaces/IPoolAddressesProvider.sol';
+
+/**
+ * @title  August 18, 2023 Spark Ethereum Proposal - Update to Aave 3.0.2, Unfreeze sDAI Market, Update DAI Interest Rate Strategy
+ * @author Phoenix Labs
+ * @dev    This proposal upgrades the pool contract implementation to Aave 3.0.2
+ * Forum:  https://forum.makerdao.com/t/phoenix-labs-proposed-changes-for-spark-for-august-18th-spell/21612
```
Comment (lucas-manuel) [inline src/proposals/20230816/SparkEthereum_20230816.sol:12]: Correct link

File: src/proposals/20230816/SparkGoerli_20230816.sol [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import { SparkPayloadGoerli, IEngine, Rates, EngineFlags } from '../../SparkPayloadGoerli.sol';
+
+import { IPoolAddressesProvider } from 'aave-v3-core/contracts/interfaces/IPoolAddressesProvider.sol';
+
+/**
+ * @title  August 18, 2023 Spark Goerli Proposal - Update to Aave 3.0.2, Unfreeze sDAI Market, Update DAI Interest Rate Strategy
+ * @author Phoenix Labs
+ * @dev    This proposal upgrades the pool contract implementation to Aave 3.0.2
+ * Forum:  https://forum.makerdao.com/t/phoenix-labs-proposed-changes-for-spark-for-august-18th-spell/21612
```
Comment (lucas-manuel) [inline src/proposals/20230816/SparkGoerli_20230816.sol:12]: Correct link

File: src/proposals/20230816/SparkGoerli_20230816.sol [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * Vote:   N/A
+ */
+contract SparkGoerli_20230816 is SparkPayloadGoerli {
+
+    address public constant POOL_ADDRESS_PROVIDER      = 0x026a5B6114431d8F3eF2fA0E1B2EDdDccA9c540E;
```
Comment (lucas-manuel) [inline src/proposals/20230816/SparkGoerli_20230816.sol:17]: Addresses are all correct

File: src/proposals/20230816/SparkEthereum_20230816.t.sol [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                baseRateConversion: 1e27,
+                borrowSpread:       0,
+                supplySpread:       0,
+                maxRate:            0.75e27,
+                performanceBonus:   0
+            })
+        );
+        assertEq(_findReserveConfigBySymbol(configsAfter, 'sDAI').isFrozen, false);
+        vm.prank(address(poolAddressesProvider));
+        assertEq(BaseAdminUpgradeabilityProxy(payable(address(pool))).implementation(), POOL_IMPLEMENTATION_NEW);
```
Comment (lucas-manuel) [inline src/proposals/20230816/SparkEthereum_20230816.t.sol:100]: No diff between tests, tests assert all desired state

File: diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+| isFrozen | true | false |
+
+
+## Raw diff
+
+```json
+{
+  "poolConfig": {
+    "poolImpl": {
+      "from": "0x62DA45546A0F87b23941FFE5CA22f9D2A8fa7DF3",
```
Comment (lucas-manuel) [inline diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md:25]: Matches primary-latest.json

File: src/proposals/20230816/SparkEthereum_20230816.t.sol [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    bytes32 public constant SPARK_ILK = "DIRECT-SPARK-DAI";
+
+    IPool public constant POOL = IPool(0xC13e21B648A5Ee794902342038FF3aDAB66BE987);
+
+    constructor() {
+        id = '20230816';
+    }
+
+    function setUp() public {
+        vm.createSelectFork(getChain('mainnet').rpcUrl, 17892780);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20230816/SparkEthereum_20230816.t.sol:37]: done

File: src/proposals/20230816/SparkGoerli_20230816.t.sol [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    bytes32 public constant SPARK_ILK = "DIRECT-SPARK-DAI";
+
+    IPool public constant POOL = IPool(0x26ca51Af4506DE7a6f0785D20CD776081a05fF6d);
+
+    constructor() {
+        id = '20230816';
+    }
+
+    function setUp() public {
+        vm.createSelectFork(getChain('goerli').rpcUrl, 9500235);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20230816/SparkGoerli_20230816.t.sol:37]: done

File: diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      "to": "0x8115366Ca7Cf280a760f0bC0F6Db3026e2437115"
```
Comment (lucas-manuel) [inline diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md:26]: Matches primary-pool-latest.json poolImpl

File: diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  },
+  "reserves": {
+    "0x6B175474E89094C44Da98b954EedeAC495271d0F": {
+      "interestRateStrategy": {
+        "from": "0x191E97623B1733369290ee5d018d0B068bc0400D",
```
Comment (lucas-manuel) [inline diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md:32]: Matches 0802 spell

File: diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        "to": "0x9f9782880dd952F067Cad97B8503b0A3ac0fb21d"
```
Comment (lucas-manuel) [inline diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md:33]: Matches 0525 spell

File: diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    },
+    "0x83F20F44975D03b1b09e64809B757c47f942BEeA": {
```
Comment (lucas-manuel) [inline diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md:36]: Confirmed sDAI

File: diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+      "isFrozen": {
+        "from": true,
+        "to": false
+      }
+    }
+  },
+  "strategies": {
+    "0x9f9782880dd952F067Cad97B8503b0A3ac0fb21d": {
```
Comment (lucas-manuel) [inline diffs/20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-pre_20230816-Ethereum-0xC13e21B648A5Ee794902342038FF3aDAB66BE987-post.md:44]: Matches 0525 spell

File: src/proposals/20230816/SparkGoerli_20230816.t.sol [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    address public constant POOL_IMPLEMENTATION_OLD = 0xF1E57711Eb5F897b415de1aEFCB64d9BAe58D312;
+    address public constant POOL_IMPLEMENTATION_NEW = 0xe7EA57b22D5F496BF9Ca50a7830547b704Ecb91F;
+
+    constructor() {
+        id = '20230816';
+    }
+
+    function setUp() public {
+        vm.createSelectFork(getChain('goerli').rpcUrl, 9500235);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/proposals/20230816/SparkGoerli_20230816.t.sol]: done

File: src/proposals/20230816/SparkGoerli_20230816.sol [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    address public constant POOL_IMPLEMENTATION        = 0xe7EA57b22D5F496BF9Ca50a7830547b704Ecb91F;
+    address public constant DAI                        = 0x11fE4B6AE13d2a6055C8D9cF65c55bac32B5d844;
+    address public constant sDAI                       = 0xD8134205b0328F5676aaeFb3B2a0DC15f4029d8C;
+    address public constant DAI_INTEREST_RATE_STRATEGY = 0x70659BcA22A2a8BB324A526a8BB919185d3ecEBC;
```
Comment (amusingaxl) [inline src/proposals/20230816/SparkGoerli_20230816.sol:21]: To validate the source code on Goerli:

1. Installed `sourcify` locally as instructed in the README
2. Edited `sourcify` env config to use valid Goerli and Mainnet nodes
3. Ran `sourcify` server
4. Validated source code through `sourcify` by running:
    ```bash
    ./validate-deploy.sh script/output/5/broadcast/primary-pool-20230714.json
    ```

Results:
```
Submitting verification for [BorrowLogic] "0x512AFEDCF6696d9707dCFECD4bdc73e9902e3c6A".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:37.517Z
Start verifying contract `0x9d0053198fc830a68165d85d4b3d33d7c002b7fc` deployed on goerli

Submitting verification for [BridgeLogic] "0x9D0053198fc830a68165d85D4b3D33D7C002B7FC".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:38.833Z
Start verifying contract `0xe25f940d74094dc0397f22f6ad6ae656f4e6297b` deployed on goerli

Submitting verification for [EModeLogic] "0xE25f940D74094dc0397F22f6aD6AE656f4e6297b".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:40.323Z
Start verifying contract `0xb5dad3aea0d92417f100655d1e9b31fab85304a3` deployed on goerli

Submitting verification for [FlashLoanLogic] "0xb5DaD3AeA0D92417f100655d1e9B31fab85304A3".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:42.287Z
Start verifying contract `0x83512f8c629da31b430f5bf68e8ee97322b90c60` deployed on goerli

Submitting verification for [LiquidationLogic] "0x83512f8C629dA31B430F5bF68e8Ee97322b90c60".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:44.463Z
Start verifying contract `0x8e343af9223733d2d41f0e66e7600033f0f996e6` deployed on goerli

Submitting verification for [PoolLogic] "0x8E343aF9223733d2d41F0e66E7600033F0f996E6".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:45.793Z
Start verifying contract `0x6264de350020ec942213c7742075096680a3a694` deployed on goerli

Submitting verification for [SupplyLogic] "0x6264dE350020EC942213C7742075096680a3A694".
Contract source code already verified. Storage Timestamp: 2023-08-15T17:43:47.250Z

///////////////////////////////// NEW POOL IMPLEMENTATION VERIFICATION //////////////////////////////
Start verifying contract `0xe7ea57b22d5f496bf9ca50a7830547b704ecb91f` deployed on goerli

Submitting verification for [Pool] "0xe7EA57b22D5F496BF9Ca50a7830547b704Ecb91F".
Contract successfully verified
```

File: src/proposals/20230816/SparkEthereum_20230816.sol [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ * Vote:   N/A
+ */
+contract SparkEthereum_20230816 is SparkPayloadEthereum {
+
+    address public constant POOL_ADDRESS_PROVIDER      = 0x02C3eA4e34C0cBd694D2adFa2c690EECbC1793eE;
+    address public constant POOL_IMPLEMENTATION        = 0x8115366Ca7Cf280a760f0bC0F6Db3026e2437115;
+    address public constant DAI                        = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address public constant sDAI                       = 0x83F20F44975D03b1b09e64809B757c47f942BEeA;
+    address public constant DAI_INTEREST_RATE_STRATEGY = 0x9f9782880dd952F067Cad97B8503b0A3ac0fb21d;
```
Comment (amusingaxl) [inline src/proposals/20230816/SparkEthereum_20230816.sol:21]: All addresses except for `POOL_IMPLEMENTATION` match this [file](https://github.com/marsfoundation/sparklend/blob/master/script/output/1/primary-latest.json).

@hexonaut said the file is to be updated **after** the spell is executed.

File: src/proposals/20230816/SparkGoerli_20230816.sol (hunk 15 above)
Comment (amusingaxl) [inline src/proposals/20230816/SparkGoerli_20230816.sol:21]: All addresses except for `POOL_IMPLEMENTATION` match this [file](https://github.com/marsfoundation/sparklend/blob/ca2b72af7c5fb790cc91eaca5d8d4c83fa37e74b/script/output/5/primary-latest.json).

@hexonaut said the file is to be updated **after** the spell is executed.

File: src/proposals/20230816/SparkGoerli_20230816.sol (hunk 3 above)
Comment (amusingaxl) [reply src/proposals/20230816/SparkGoerli_20230816.sol:12]: Link is correct.

File: src/proposals/20230816/SparkEthereum_20230816.sol (hunk 2 above)
Comment (amusingaxl) [reply src/proposals/20230816/SparkEthereum_20230816.sol:12]: Link is correct.

**Review by amusingaxl — APPROVED** [review]
As a MakerDAO contributor, I performed the following checks:
- [x] Source code of the deployed contracts match
- [x] Addresses (mostly) match the [main repo](https://github.com/marsfoundation/sparklend).
- [x] Tests cover the actions being taken in the spell.
- [x] Links to the proposals in MakerDAO Forum.

Things that I DID NOT check:
- Former values in spell history. I'm assuming they are correct.

**Pull request merged**